contracts/contracts/buyback/BaseBuyback.sol

```diff
@@ -0,0 +1,266 @@
+            IERC20(oToken).approve(universalRouter, 0);
+        }
+
+        universalRouter = _router;
+
+        emit UniswapUniversalRouterUpdated(_router);
+    }
+
+    /**
+     * @dev Sets the address that receives the OGV buyback rewards
+     * @param _address Address
+     */
+    function setRewardsSource(address _address) external onlyGovernor {
+        _setRewardsSource(_address);
+    }
+
+    function _setRewardsSource(address _address) internal {
+        require(_address != address(0), "Address not set");
+        rewardsSource = _address;
+        emit RewardsSourceUpdated(_address);
+    }
+
+    /**
+     * @dev Sets the address that can receive and manage the funds for Treasury
+     * @param _address Address
+     */
+    function setTreasuryManager(address _address) external onlyGovernor {
+        _setTreasuryManager(_address);
+    }
+
+    function _setTreasuryManager(address _address) internal {
+        require(_address != address(0), "Address not set");
+        treasuryManager = _address;
+        emit TreasuryManagerUpdated(_address);
+    }
+
+    /**
+     * @dev Swaps half of `oTokenAmount` to OGV
+     *      and the rest to CVX and finally lock up CVX
+     * @param oTokenAmount Amount of OUSD/OETH to swap
+     * @param minOGV Minimum OGV to receive for oTokenAmount/2
+     * @param minCVX Minimum CVX to receive for oTokenAmount/2
+     */
+    function swap(
+        uint256 oTokenAmount,
+        uint256 minOGV,
+        uint256 minCVX
+    ) external onlyGovernorOrStrategist nonReentrant {
+        _swap(oTokenAmount, minOGV, minCVX);
+    }
+
+    function _swap(
+        uint256 oTokenAmount,
+        uint256 minOGV,
+        uint256 minCVX
+    ) internal {
+        require(oTokenAmount > 0, "Invalid Swap Amount");
+        require(universalRouter != address(0), "Uniswap Router not set");
+        require(rewardsSource != address(0), "RewardsSource contract not set");
+        require(minOGV > 0, "Invalid minAmount for OGV");
+        require(minCVX > 0, "Invalid minAmount for CVX");
+
+        uint256 ogvBalanceBefore = IERC20(ogv).balanceOf(rewardsSource);
+        uint256 cvxBalanceBefore = IERC20(cvx).balanceOf(address(this));
+
+        // Build swap input
+        bytes[] memory inputs = new bytes[](2);
+
+        inputs[0] = abi.encode(
+            // Send swapped OGV directly to RewardsSource contract
+            rewardsSource,
+            oTokenAmount / 2,
+            minOGV,
+            _getSwapPath(ogv),
+            false
+        );
+
+        inputs[1] = abi.encode(
+            // Buyback contract receives the CVX to lock it on
+            // behalf of Strategist after the swap
+            address(this),
+            oTokenAmount / 2,
+            minCVX,
+            _getSwapPath(cvx),
+            false
+        );
+
+        // Transfer OToken to UniversalRouter for swapping
+        IERC20(oToken).safeTransfer(universalRouter, oTokenAmount);
```

Comment:
We don't have to use `safeTransfer` for our own tokens, since they are ERC20 compatible.

Reply:
Done as well